README.md
# node-red-contrib-iiot-rpi-ads1115

[![platform](https://img.shields.io/badge/platform-Node--RED-red)](https://nodered.org)
[![platform](https://img.shields.io/badge/platform-Raspberry--Pi-ff69b4)](https://www.raspberrypi.com/)
![NPM version](https://badge.fury.io/js/node-red-contrib-iiot-rpi-ads1115.svg)
![NPM](https://img.shields.io/npm/l/node-red-contrib-iiot-rpi-ads1115)
[![Donate](https://img.shields.io/badge/Donate-PayPal-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ZDRCZBQFWV3A6)

A Node-Red node for 16bit ads1115 analog to digital converter<br>

![image info](images/ads1115.png)

Example  **ads1115** hat from [joy-it](https://www.joy-it.net/en/products/RB-ADC01)

![image info](images/ads1115module.png)

Example **ads1115** module

## Installation
Install with Node-Red Palette Manager or npm command:
```
cd ~/.node-red
npm install node-red-contrib-iiot-rpi-ads1115
```

## Usage
- This node works on Raspberry Pi with 32bit or 64bit OS.
- The output value is mV or adc raw data.
- For multi node, all values send as array.
- Enable I2C with raspi-config.

## Array Index on Multi Node

|Input|Array-Index
|:----|:---:|
|A0|0|
|A1|1|
|A2|2|
|A3|3|

[CHANGELOG](CHANGELOG.md)<br>

CHANGELOG.md
1.0.20
- Initial Release

1.0.21
- Add node for multiple channel read and send as array


ads1115-m.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
	const syslib = require("./lib/syslib.js");
	const sysmodule = syslib.LoadModule("rpi_ads1115");

    RED.nodes.registerType("ads1115-m", function(n) {
		var node = this;
		RED.nodes.createNode(node, n);

		node.devadr = parseInt(n.devadr);
		node.rate = parseInt(n.rate);
		node.gain0 = parseInt(n.gain0);
		node.gain1 = parseInt(n.gain1);
		node.gain2 = parseInt(n.gain2);
		node.gain3 = parseInt(n.gain3);
		node.tupdate = parseInt(n.tupdate);
		node.onchange = n.onchange;
		node.rawdata = n.rawdata;

		node.name = n.name || "ads1115-m" + " @" + (0x48 + parseInt(this.devadr)).toString(16).toUpperCase();

		if (sysmodule === undefined)
			syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
		else if (!sysmodule.open(node.devadr))
			syslib.outError(node, "open error", "i2c port not open, check i2c");
		else
			node.id_tupdate = setInterval(function () {
				var readval = sysmodule.readmulti(node.devadr, node.rate, node.gain0, node.gain1, node.gain2, node.gain3, node.rawdata);

				if (readval === undefined)
					syslib.outError(node, "read error", "device not read, check i2c and device");
				else if (!node.onchange || (JSON.stringify(node.readval) !== JSON.stringify(readval))) {
					syslib.outText(node, JSON.stringify(readval));
					node.send({ payload: readval });
					node.readval = readval;
				}
			}, node.tupdate);
		
		node.on('close', function () {
			clearInterval(node.id_tupdate);
		});
	});
}


lib/syslib.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use node file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

const fs = require('fs');

module.exports.outError = function (node, errShort, errLong) {
	if (node.save_txt === errShort)
		return true;

	node.save_txt = errShort;
	node.status({ fill: "red", shape: "ring", text: errShort });
	node.error(errLong);

	return true;
}

module.exports.outText = function (node, txt) {
	if (node.save_txt === txt)
		return;

	node.save_txt = txt;
	node.status({ fill: "blue", shape: "ring", text: txt });
}

function getRpiTarget()
{
	if (process.platform !== 'linux')
		return undefined;
	
	var cpuinfo = fs.readFileSync("/proc/cpuinfo").toString();

	if (cpuinfo.indexOf(": BCM") === -1)
		return undefined;
	
	if (process.arch === "arm64")
		return "_s64.node";
	else if (process.arch === "arm")
		return "_s32.node";
	
	return undefined;
}

module.exports.LoadModule = function (module) {
	try {
		var modName = getRpiTarget();
		if (modName !== undefined)
			return require("./" + module + modName);
	}
	catch (e) { console.log(e); }
	return undefined;
}


ads1115.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
	const syslib = require("./lib/syslib.js");
	const sysmodule = syslib.LoadModule("rpi_ads1115");

    RED.nodes.registerType("ads1115", function(n) {
		var node = this;
		RED.nodes.createNode(node, n);

		node.devadr = parseInt(n.devadr);
		node.mux = parseInt(n.mux);
		node.rate = parseInt(n.rate);
		node.gain = parseInt(n.gain);
		node.tupdate = parseInt(n.tupdate);
		node.onchange = n.onchange;
		node.rawdata = n.rawdata;

		var inputname;
		switch(node.mux) {
			case 0: inputname = "0-1"; break;
			case 1: inputname = "0-3"; break;
			case 2: inputname = "1-3"; break;
			case 3: inputname = "2-3"; break;
			case 4: inputname = "0-G"; break;
			case 5: inputname = "1-G"; break;
			case 6: inputname = "2-G"; break;
			case 7: inputname = "3-G"; break;
		}
		node.name = n.name || "ads1115" + " @" + (0x48 + parseInt(this.devadr)).toString(16).toUpperCase() + "#" + inputname;

		if (sysmodule === undefined)
			syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
		else if (!sysmodule.open(node.devadr))
			syslib.outError(node, "open error", "i2c port not open, check i2c");
		else
			node.id_tupdate = setInterval(function () {
				var readval = sysmodule.read(node.devadr, node.mux, node.rate, node.gain, node.rawdata);

				if (readval === undefined)
					syslib.outError(node, "read error", "device not read, check i2c and device");
				else if (!node.onchange || (readval !== node.readval)){
					syslib.outText(node, readval);
					node.send({ payload: readval });
					node.readval = readval;
				}
			}, node.tupdate);
		
		node.on('close', function () {
			clearInterval(node.id_tupdate);
		});
	});
}
